[GStreamer] triggerRepaint() should allow a null previousBuffer

eocanha opened the pull request:
#### 29aaf6c0d8ef780c2dc65979da05607ecc1dd003
<pre>
[GStreamer] triggerRepaint() should allow a null previousBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=291277">https://bugs.webkit.org/show_bug.cgi?id=291277</a>

Reviewed by Philippe Normand.

We&apos;re getting some crashes downstream in
<a href="https://ytlr-cert.appspot.com/2021/main.html?&amp">https://ytlr-cert.appspot.com/2021/main.html?&amp</a>;test_type=progressive-test&amp;command=run
on Raspberry Pi, and they&apos;re caused by RELEASE_ASSERT(previousBuffer) in
MediaPlayerPrivateGStreamer::triggerRepaint(). When debugged, I
realized that the null previousBuffer was set by gst_buffer_copy_deep()
in MediaPlayerPrivateGStreamer::flushCurrentBuffer(). This means that
the deep copy failed to produce a valid buffer. This is sort of expected
in Raspberry Pi when using the OMX decoder, since those buffers come
from a special buffer pool managed by OMX and reference scarce GPU
memory (caps: video/x-raw(memory:GLMemory)).

This patch removes the assert and adds an extra log line to make future
developers debugging issues aware of the fact that null buffers are
being used now and why.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::triggerRepaint): Remove assert.
(WebCore::MediaPlayerPrivateGStreamer::flushCurrentBuffer): Warn that the buffer couldn&apos;t be copied, if that&apos;s the case.

Canonical link: <a href="https://commits.webkit.org/293459@main">https://commits.webkit.org/293459@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f6dfc226129548e83414c82a22583926787972d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/21254 "Build was cancelled. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Checked out pull request; Running add-reviewer-to-commit-message; Compiled WebKit (warnings); layout-tests (exception)") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3741,7 +3741,8 @@ void MediaPlayerPrivateGStreamer::triggerRepaint(GRefPtr<GstSample>&& sample)
             auto previousBuffer = gst_sample_get_buffer(m_sample.get());
-            RELEASE_ASSERT(previousBuffer);
+            // This is not an assert because on some embedded platforms the buffer can't be deep copied by flushCurrentBuffer().
+            GST_DEBUG("previousBuffer is null");
```
I'm not sure this GST_DEBUG is much useful

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3843,6 +3844,8 @@ void MediaPlayerPrivateGStreamer::flushCurrentBuffer()
         const GstStructure* info = gst_sample_get_info(m_sample.get());
         auto buffer = adoptGRef(gst_buffer_copy_deep(gst_sample_get_buffer(m_sample.get())));
+        if (!buffer)
+            GST_DEBUG("Buffer couldn't be deep-copied on this platform, setting null buffer on the sample instead");
```
GST_DEBUG_OBJECT here please

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -25,6 +25,7 @@
 
 #include "config.h"
 #include "MediaPlayerPrivateGStreamer.h"
+#include <gst/gstinfo.h>
```
This was inserted by LSP I guess? :) Not needed

eocanha replied:
Argh! 🤦 Thanks! 🙏

eocanha committed:
[GStreamer] triggerRepaint() should allow a null previousBuffer
https://bugs.webkit.org/show_bug.cgi?id=291277

Reviewed by Philippe Normand.

We're getting some crashes downstream in
https://ytlr-cert.appspot.com/2021/main.html?&test_type=progressive-test&command=run
on Raspberry Pi, and they're caused by RELEASE_ASSERT(previousBuffer) in
MediaPlayerPrivateGStreamer::triggerRepaint(). When debugged, I
realized that the null previousBuffer was set by gst_buffer_copy_deep()
in MediaPlayerPrivateGStreamer::flushCurrentBuffer(). This means that
the deep copy failed to produce a valid buffer. This is sort of expected
in Raspberry Pi when using the OMX decoder, since those buffers come
from a special buffer pool managed by OMX and reference scarce GPU
memory (caps: video/x-raw(memory:GLMemory)).

This patch removes the assert and adds an extra log line to make future
developers debugging issues aware of the fact that null buffers are
being used now and why.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::triggerRepaint): Remove assert.
(WebCore::MediaPlayerPrivateGStreamer::flushCurrentBuffer): Warn that the buffer couldn't be copied, if that's the case.

Canonical link: https://commits.webkit.org/293459@main
